xunit-autolabeler-v2/ast_parser/python/test_parser.py

```diff
@@ -0,0 +1,227 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import ast
+import os
+import sys
+from typing import Any, Dict, List, Tuple
+
+
+from ast_parser.core import constants
+from . import drift_test
+
+
+def _get_test_nodes(parsed_nodes: List[Any]) -> List[Any]:
+    """Helper function to extract test methods wrapped in classes
+
+    Args:
+        parsed_nodes (List[ast.AST]): a list of nodes parsed
+                                      from a Python test file
+
+    Returns:
+        List[ast.AST]: a list of possible test nodes
+    """
+
+    possible_test_nodes = []
+    for node in parsed_nodes:
+        possible_test_nodes.append(node)
+        if 'ClassDef' in str(type(node)):
+            # Concatenate node.body (which is an array)
+            possible_test_nodes += node.body
+
+    return [node for node in possible_test_nodes
+            if hasattr(node, 'name') and node.name.startswith('test_')]
+
+
+def get_test_methods(test_path: str) -> List[Any]:
+    """Gets the top-level methods within a test file
+
+    Args:
+        source_path: path to the file to process
+
+    Returns:
+        List[ast.AST]: a list of the top-level
+                       methods within the provided file
+    """
+    try:
+        with open(test_path, 'r') as file:
+            content = ''.join(file.readlines())
+            parsed_nodes = list(ast.iter_child_nodes(ast.parse(content)))
+
+            test_nodes = _get_test_nodes(parsed_nodes)
+
+            for node in test_nodes:
+                node.test_path = test_path
+
+            # Verify file contains no duplicate method names
+            # (Only relevant for test methods wrapped in classes)
+            used_test_names = set()
+            for node in test_nodes:
+                if node.name in used_test_names:
+                    raise ValueError(
+                        f'Test name {node.name} in file'
+                        f' {test_path} must be unique.')
+                used_test_names.add(node.name)
+
+            return test_nodes
+    except IOError as err:
+        # Fail gracefully if a file can't be read
+        # (This shouldn't happen, but if it doess
+        #  we don't want to 'break the build'.)
+        sys.stderr.write(
+            f'WARNING: could not read file: {test_path}\n')
+        sys.stderr.write(
+            f'\t{str(err)}\n')
+
+        return []
+
+
+def get_test_key_to_snippet_map(
+    test_methods: List[Any]
+) -> Dict[Tuple[str, str], Tuple[str, str]]:
```

Comment:
Here and there, the return value is a dict. What's the value? I only know it's `test data`, but what's inside?

Reply:
Done.